### DISCUSSION TITLE
freetz-ng on 7590AX_08.00 not supporting /dev/ttyUSB0 with ch341 device driver

### DISCUSSION THREAD

**Discussion by Ircama** [Freetz-NG/freetz-ng#1004] [Allgemein]
Tested latest *freetz-ng* pull on a FRITZ!Box 7590 AX (Fritz_Box_HW259) with FRITZ!OS 07.90: for possible kernel incompatibility, a USB peripheral device provided with a ch34X interface is not seen by the OS (/dev/ttyUSB0 is missing). `lsusb-freetz -t` lists the port but does not fill the driver label (`Driver=` instead of `Driver=ch341`).

What works:

- the /dev/ttyUSB0 device driver of the latest *freetz-ng* is available and works well on a FRITZ!Box 7590 AX (Fritz_Box_HW259) with FRITZ!OS 07.80 and also 07.81 (max speed 115kbps).
- `./tools/push_firmware` works flawlessly with Fritz_Box_HW259 - FRITZ!Box 7590 AX v2 Edition International

**Comment by fda77** [comment]
maybe you need latest kernel sources from AVM for this labor.
but avm gives a fuck on GPL and does not publish gpl-sources for labor
You could try to use the kernel module from the release...

**Comment by Ircama** [comment]
Unfortunately, also with FRITZ!OS 8 the USB serial devices are not supported.

- /dev/ttyUSB0 is missing
- `lsusb-freetz -vt` lists the port (so the kernel detects the USB device) but does not fill the driver label (Driver= instead of Driver=ch341):

```
# lsusb-freetz
...
Bus 003 Device 002: ID 1a86:7523 QinHeng Electronics CH340 serial converter
...

# lsusb-freetz -vt
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Vendor Specific Class, Driver=, 12M
```

*ch341.ko* and *usbserial.ko* are selected in `make menuconfig`.

With FRITZ!OS 8 `uname -a` reports again the old 4.9.325 release:

```
Linux fritz.box 4.9.325 #1 SMP 2024-09-03 mips GNU/Linux
```

Any hint? Thanks

**Reply by Ircama** [reply]
This is missing in the `dmesg` of 08.00:

```
[   61.199758] [module-mem] module 'usbserial' reserved size 40960 is to great for demand size 32768 - waste 8192 (module_alloc_waste=192512)
[   61.199792] [module-mem] give 0x8000 bytes at 983c7000 to module 'usbserial' (0x161000 bytes left)
[   61.203523] usbcore: registered new interface driver usbserial
[   61.220516] usbcore: registered new interface driver ch341
[   61.220707] usbserial: USB Serial support registered for ch341-uart
[   61.221362] ch341-uart ttyUSB0: no of_node; not parsing pinctrl DT
[   61.224799] usb 3-1: ch341-uart converter now attached to ttyUSB0
```

In any case, 7.81 has a poor /dev/ttyUSB0 USB serial support, with max rate 115200 baud.

**Comment by fda77** [comment]
The driver is "missing" because the avm kernel sources are not building.
I disabled it now: 6ac7316b7d51495f9c2cc44a02397cbef473f7c7
At the moment i do not have time to search what avm changed and to make it work again - maybe you have ???
Same for 7530ax src: 7bf3463207774428e3c296c75bad5db60f63a666

**Reply by Ircama** [reply]
Thanks for your swift explanation @fda77 and for your work to maintain this repository. I think I understood that the driver is missing because the build fails (hence the need to disable it). I would actually like to be able to restore it because I have almost finished developing a Python software for freetz-ng that manages potentially all BLE home sensors (I was going to publish the code as soon as the Web GUI is completed); however, it is based on BLE receivers interfaced to the device via USB and I realize that this software will not work with the new versions of FRITZ!OS.

I would have a little bit of time but I cannot find the failing code myself, I do not know how to reproduce the compilation error and in general I admit that I have no experience in kernel programming.

**Comment by fda77** [comment]
If a device (Fritzbox) has device drivers (kernel module) depends on the device and firmware version. 
Every time the changes need to be fixed so it compiles.  eg for https://freetz-ng.github.io/freetz-ng/CHANGELOG.html#ng23080 19 new "AVM sources" where added, meanwhile 94 different are in freetz!
I have no time for all of these

**Reply by fda77** [reply]
not working 7530ax+7590ax 8.0 are now marked: 37af062a08cf566651b3ca5b113295ba37fabc6f

**Reply by Ircama** [reply]
All works now  —  thanks for the support!

I originally assumed that `usbserial` support with `ch341` was dropped starting with FRITZ!OS 8.x. However, after testing on a 7590 AX (AVME) with **FRITZ!OS 08.02**, everything works as expected.

That said, the issue actually arises when dealing with **Labor/Beta firmware versions**, such as **FRITZ!OS 08.10 Labor** at the moment. In these cases, AVM does not provide enough information to build the necessary kernel modules.

For instance, after selecting now version **FRITZ!OS 08.10 Labor** with the target 7590 AX AVME, I get the `make menuconfig` message

```
Kernel modules (not available, ask fritzbox_info@avm.de)
```

If trying by curiosity to overcome it, attempting to enable `ch341` via `make kernel-menuconfig`, the system outputs:

```
Freetz-NG 25925M-031b31adf9 master 2025-05-03
JUIS: Using cached value ... No cached value, asking avm for latest firmware ... no valid answer, trying backup ...
make[1]: *** [make/include/700-image.mk:156: dl/fw/7590_AX-xx_labor.detected.image] Terminated
make: *** [Makefile:49: envira] Terminated
```

So the limitation is not in the stable FRITZ!OS version, but rather in AVM’s incomplete metadata for beta firmware images.

**Testing procedure with stable FRITZ!OS version:**

```bash
rm -rf freetz-ng  # ~10 GB needed
git clone https://github.com/Freetz-NG/freetz-ng
cd freetz-ng
make menuconfig
# - Set user level to: Expert mode (to enable Kernel modules)
# - Select hardware: 7590-AX, AVME
# - Select firmware: FRITZ!OS 08.00+
# - Kernel modules > drivers > ch341.ko

make
# *** Please re-run
make
./tools/push_firmware
```

**Reply by fda77** [reply]
> Kernel modules (not available, ask fritzbox_info@avm.de)

best way to get avm-patches for the kernel asap is to write an email! (i NEVER do this, its YOUR device)
